xlet settings: add list widget

collinss opened the pull request:
* Allows adding/removing/editing/reordering entries in settings
* Allows a variable number of columns, each with it's own properties and data type
* Currently supports string, integer, float, bool, sound, icon, keybinding, and file variable types (more can be added easily - let me know if you have any ideas or requests)

![list widget 2](https://cloud.githubusercontent.com/assets/4095021/23190301/0e9be700-f865-11e6-851c-e9c3c7617590.png)
![list widget 3](https://cloud.githubusercontent.com/assets/4095021/23190302/0eb5bc98-f865-11e6-880a-a724ea57adf8.png)

Odyseus commented:
Hello, @collinss.

I think that this is amazing! It isn't just useful, but absolutely needed!

~~I already can think of an applet that could make use of this ([Feeds Reader](https://github.com/linuxmint/cinnamon-spices-applets/tree/master/feeds%40jonbrettdev.wordpress.com)). It uses a 400 lines long Python script to create the widget to manage its feeds.~~ <- (You already mentioned [here](https://github.com/linuxmint/cinnamon-spices-applets/pull/154#issue-209815857)) In one of my applets, I could convert an entire setting page that holds more than 40 preferences of different types into just one list widget!

#### Suggestions/Feedback/Questions

1. ~~The **file** type can only select directories, not files.~~ But looking more closely, that can be changed in the settings-schema.json file (`"select-dir": true/false`). BTW, in the current settings-schema.json file that parameter is set as `"select-dir": 100` (an integer instead of a boolean).
2. Would it be possible to make all the entries in the list in-line editable? Not a must have, but it would facilitate the edition of large amount of data.
3. I would really like to also see available **iconchooser** and **keybinding** elements if possible.
4. I thing that all text boxes should be flexible. Maybe add the existent "expand-width" parameter that all entry types already have?
5. I added two long paths and the entire window resized horizontally beyond the limits of the screen and without the possibility to be resized nor maximized. I think that to fix this, the columns could be resizable and its content *ellipsized*.

collinss commented:
@Odyseus Thanks for the feedback, and for breaking my pr for me (no really, it helps).

1. You can leave the select-dir off and it will default to file select rather than directory. Also, ```"select-dir": 100``` was fixed as well (nice catch).
2. I think this might turn out to be not only be difficult to implement, but a bad idea as well. Certain items (eg. keybindings) have particular formatting, and an incorrect format could spell disaster.
3. Added. I also added soundchooser as well because I could and it was easy. I was going to add datechooser, but there are some challenges since the date is stored as a json style object rather than a basic type. I might still if I can figure out a good way to make it play nice with the Gtk tree widgets without changing the way datechoosers work.
4. It should be up to the applet dev to determine if it's needed or not. The "expand-width" property is already available. The individual authors should be the ones to decide if a particular field needs to be larger or not.
5. Fixed.

collinss commented:
I just updated the add/edit dialog to look a little more like the rest of settings:
![list widget 4](https://cloud.githubusercontent.com/assets/4095021/24023886/97d9bae2-0a85-11e7-9acc-512f03dc795e.png)

collinss commented:
I think this is ready.

Odyseus commented:
Hello, @collinss

I tested your last changes and all seems to be working perfect.

> It should be up to the applet dev to determine if it's needed or not. The "expand-width" property is already available. The individual authors should be the ones to decide if a particular field needs to be larger or not.

I think that you misunderstood my request on point 4. I was talking about the entries on the following dialog:

![testlistwidget2](https://cloud.githubusercontent.com/assets/3822556/24080155/db4fb10c-0c77-11e7-9947-8e32c853b05d.png)

See those highlighted entries? Those are the ones that should be able to expand. Right now, setting the *expand-width* property on *string* or *icon* column types does not make those entries expandable.

There is another thing that I'm curious about (not directly related to this pull request). Why the change and use of URIs (file:///path/to/file/or/folder) on *filechooser* elements, and now on the list widget's *file* type columns, when in the past the *filechooser* elements returned a path (/path/to/file/or/folder)? That is kind of inconsistent, especially considering that the *iconchooser* elements are still returning a path on the new settings system, just like the old settings system does. This change caused me a real headache when I first started using the new settings system, because the function I used was expecting a path, not an URI.

collinss commented:
Ah, I see what you mean. I didn't realize those didn't support the expand property (someone else added that after-the-fact - I had nothing to do with it). I'll see about adding them, but I'll probably do so in a separate pr as that's kind of outside the scope of this one. As for the uri/path issue, when I did the revamp of xlet settings, I basically used the gsettings version of the widgets if they already existed, and only ported from the applet settings if they didn't. Most likely there was already a filechooser, and I didn't realize the difference.

collinss commented:
@Odyseus Apparently I misunderstood you again. I just updated it again, and it should be working now. Let me know if it doesn't.

Odyseus commented:
Hello, @collinss.

Now we are cooking! This is how it looks now with the *expand-width* option set:

![testlistwidget3](https://cloud.githubusercontent.com/assets/3822556/24116331/3a4fa4ba-0d85-11e7-938e-e5820cc7d257.png)

Yes, I know. It doesn't look pretty. But at least they are comfortable to work with. That's why [I added the *expand-width* option](https://github.com/linuxmint/Cinnamon/pull/5891) to *entry* and *iconfilechooser* elements (I'm the culprit of that. LOL).

By the way, thanks again for adding this new list widget. This adds so much more possibilities and adds a most needed *power* to Cinnamon's native settings system.

Driglu4it commented:
Hello, i have a question. How use a list widget in applet (how to get title of column and use it in applet)?
Sorry, for my bad English...
